### PR TITLE
Task#60624 Sign in via google+

### PR DESCRIPTION
**Pull request by ngocvnn**
Description:
https://dev.framgia.com/redmine/issues/60624

### CODE REVIEW

File: app/scripts/factories/account.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/* global angular, Firebase */
+
+(function(){
+  ' use strict';
```
Comment (xuanduc987) [inline app/scripts/factories/account.js]: remove leading space

File: app/views/nav.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        <ul class="nav navbar-nav">
+            <li class="nav-item active">
+                <a class="nav-link" href="#">Home <span class="sr-only">(current)</span></a>
+            </li>
+            <li class="nav-item">
+                <a class="nav-link" href="#">Features</a>
+            </li>
+           
+        </ul>
+        <div class="pull-right">
```
Comment (xuanduc987) [inline app/views/nav.html]: Need to show login link

File: app/scripts/factories/account.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+       .module('app.factories')
+       .factory('AccountFactory', AccountFactory);
+        
+        function AccountFactory () {
+            var ref = new Firebase('https://cbsfr.firebaseio.com/users');
+            var account = {};
+           
+            account.login = function () {
+                ref.authWithOAuthPopup("google", {scope: 'email'}).then(function(authData) {
+                    window.location.href = "/";
```
Comment (xuanduc987) [inline app/scripts/factories/account.js]: use `$location`

File: app/scripts/factories/account.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                    console.log(authData.google.email);
+                }).catch (function (error) {
+                    console.error("Authentication failed google:", error); 
+                });
+            }
+           
+            account.logout = function () {
+                ref.unauth();
+                window.location.href="/login";
```
Comment (xuanduc987) [inline app/scripts/factories/account.js]: same as above

File: app/scripts/factories/account.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/* global angular, Firebase */
+
+(function(){
+  ' use strict';
+    angular
+       .module('app.factories')
+       .factory('AccountFactory', AccountFactory);
+        
+        function AccountFactory () {
+            var ref = new Firebase('https://cbsfr.firebaseio.com/users');
```
Comment (xuanduc987) [inline app/scripts/factories/account.js]: DataService('users')

File: app/scripts/config.js [hunk 6]
```diff
@@ -4,7 +4,7 @@
   'use strict';
   angular.module('app.config')
     .constant('VERSION', '0.0.1')
-    .constant('UPSTREAM', 'https://fepas-cbs-dev.firebaseio.com')
+    .constant('UPSTREAM', 'https://cbsfr.firebaseio.com')
```
Comment (xuanduc987) [inline app/scripts/config.js]: Don't change this

File: app/scripts/controllers/login_controller.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+/* global angular */
+
+(function(){
+    'use strict';
+    angular
+        .module('app.controllers')
+        .controller('LoginController', function(AccountFactory){
```
Comment (xuanduc987) [inline app/scripts/controllers/login_controller.js]:
``` javascript
angular
  .module('app.controllers')
  .controller('LoginController', LoginController);

LoginController.$inject = ['AccountFactory'];

function LoginController(AccountFactory) {
  ...
};
```

File: app/scripts/controllers/nav_controller.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/* global angular */
+
+(function(){
+    'use strict';
+    angular
+        .module('app.controllers')
+        .controller('NavController', function (AccountFactory) {
```
Comment (xuanduc987) [inline app/scripts/controllers/nav_controller.js]: Same as above

File: app/scripts/app.js [hunk 9]
```diff
@@ -40,12 +40,12 @@
...
   // angularRoute
   routeFunction.$inject = ['$routeProvider'];
-  function routeFunction($routeProvider){
+  function routeFunction($routeProvider) {
     $routeProvider
       .when('/', {
         templateUrl: 'views/main.html',
         controller: 'MainController',
-        controllerAs: 'main'
+        controllerAs: 'vm'
```
Comment (xuanduc987) [inline app/scripts/app.js]: Please keep it as `main`

File: app/scripts/controllers/login_controller.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        LoginController.$inject = ['$window', 'AccountFactory'];
+        
+        function LoginController($window, AccountFactory){
+            var vm = this;
+            
+            vm.inLoggedin = AccountFactory.active();
+            
+            vm.loginGoogle = function () {
+                AccountFactory.login().then(function(){
+                    $window.location.href = '/';
```
Comment (xuanduc987) [inline app/scripts/controllers/login_controller.js]: Please use `$location`

File: app/scripts/controllers/nav_controller.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+            if (vm.isLoggedin) {
+                current_user = AccountFactory.current_user();
+                vm.fullname = current_user.google.displayName;
+                vm.image_url = current_user.google.profileImageURL;
+            }
+            
+            vm.logout = function () {
+                if (AccountFactory.logout()) {
+                    $timeout(function(){
+                        $window.location.href = '/';
```
Comment (xuanduc987) [inline app/scripts/controllers/nav_controller.js]: Please use `$location`
And why `$timeout` is needed?

File: app/scripts/factories/account.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            };
+           
+            account.logout = function () {
+                auth.$unauth();
+                return true;
+            };
+            
+            account.active = function () {
+                var authData = auth.$getAuth();
+                if (authData !== null) {return true;}
```
Comment (xuanduc987) [inline app/scripts/factories/account.js]: `return authData !== null`

**Comment by xuanduc987** [comment]
:ok:

**Comment by yeuem1vannam** [comment]
⭕ K

**Pull request merged**